### DISCUSSION TITLE
Build for GCW0

### DISCUSSION THREAD

**Discussion by 8Gilles8** [grrk-bzzt/kq-fork#157] [Q&A]
Hi,

I've recently compile a GCW0 version of kq-fork for my RG350M. I'm so glad I've found this game. I never got to play it on my GP2X years ago.
I'd like to make the GCW0 version available there: [https://github.com/eduardofilo/RG350_adam_image/wiki/En:-8.-Compatible-game-ports](https://github.com/eduardofilo/RG350_adam_image/wiki/En:-8.-Compatible-game-ports)
Would you mind if I did?
I will include your readme in my opk release.

I still have to modify the code a little to prevent user from changing screen resolution.
I also have a bug when entering Rufus' Cabin: it crashes my RG350M and on my PC there's a minor graphical glitch. I still have to find out to fix it.
Apart from that, there is also a problem with some musics. It seems there is a problem with libmikmod. I'm waiting for a new OD Beta update which can fix it. In the meantime, I renamed the files responsible for the crash so that they won't load.

***SPOILER ALERT***
I've made a complete playthrough to see if everything was okay. 
I might have missed something, but after I got teleported to Maldea, I lost Temmin. Then, I managed to go back to the Manor and got another partner. After that, I went to the Coliseum and won. Temmin showed up and told me about Opal equipment. I thought he would go to the manor again, but he never showed up anymore. 
I'm not too familiar with lua, but I think there's something missing in `coliseum.lua` between lines 209 to 211 and also between lines 245 to 247. I don't see what is supposed to get the missing partner back to the Manor.

***END OF SPOILER ALERT***

Anyway, thanks for bringing life back into this game. I hope you'll soon be able to add what's still missing (although it is still very enjoyable the way it is now).

Gilles

**Comment by OnlineCop** [comment]
Hi Gilles,

I'm unfamiliar with GCW0, but I'm happy to see that you've been able to compile kq-fork for it.

This project is licensed as GPL v2, so feel free to clone, recompile, and repackage it as such.

I appreciate that you've made a complete playthrough. There are still some storyline elements that need to be fixed, like those that you mentioned. @teamterradactyl was one of the developers that had previously created a bunch of the new maps and tied a bunch of the story elements together, but I don't know whether she is active, and I don't know whether she had any plans for the story beyond what we have in the game currently.

I'm hoping to have some available time over the next couple of weeks. I've usually stuck with work related to the game engine itself, but I can take a stab at the Lua stuff (it's been a while since I've touched that language) to see what's going on with Temmin, and possibly bugs with any of the other party members. Thanks for bringing it to our attention!

I wish you the best with your GCW0 version, and that really comprehensive Compatible Game Ports project that you're putting together. That looks **awesome**!

-OnlineCop

**Reply by 8Gilles8** [reply]
Hi OnlineCop,

thanks. 
I forgot another bug (well, it made me rich so fixing it is not that important):
if you have more than one item of the same kind, you can sell them in bulk. If you sell all of them, the item line is still showing. You can keep selling it one at a time (it says 1 of 0).
Actually, I just tested this with a lone item and it also seems to work.
You can then buy something very expensive and keep on selling it until the button breaks down (or you just get tired of it).
It might be in `itemmenu.cpp` in `removeIndex` (or in `remove_item`?).

It reminds me of another one: when you get another caracter, it's equipment appears also in item list. But it disappears if you sort the list. It might be linked to the previous bug.


Regarding the Compatible Game Ports project, it's not mine. kq-fork would be my first contribution.
But you're right: it's great.

Gilles

**Comment by z9484** [comment]
What branch were you using? I know I looked at that coliseum code on road_to_1.0...
I got it to crash on win10 going into rufus' cabin, which is a good sign. I will try to see what's going on.

**Reply by 8Gilles8** [reply]
Hi z9484,

I used the command `git clone https://github.com/OnlineCop/kq-fork`. So, I guess it got me default branch which was road_to_1.0 at the time.

I didn't test a windows build. The crash in rufus' cabin didn't occur on Ubuntu 22.10, but there's the display glitch. I didn't have the time to investigate about it yet.

**Reply by z9484** [reply]
Interesting... https://github.com/OnlineCop/kq-fork/blob/road_to_1.0/scripts/coliseum.lua#L260 adds the other character to the manor. I tested it today while having no partner and having a partner and it worked fine for me. The old partner either joins your party(without asking) or goes back to the manor.

**Reply by 8Gilles8** [reply]
You're right. I didn't expect the call to `add_to_manor` to be after the `if` and `else`.
That's weird. The character should be back. I don't know if there's something specific to my RG30 build or if I did something that triggered a bug. 
I didn't mess with the savegame file at the time. Actually, I only tempered the file well after that to get the missing character back.
I don't know if I kept a save close enough to coliseum battle. If I have the time, I'll try to test the same savegame on RG350 and on linux PC.

Speaking of missing characters: I never found Sensar. It might be normal. Is sidequest7 fully implemented? I couldn't find the code that triggers it.

**Reply by z9484** [reply]
He's easy to miss we probably need to add some hints. In the opal cave you have to defeat 6 ghost miners and then he shows up.

**Reply by 8Gilles8** [reply]
Hi and happy new year everyone.

I wanted to investigate Temmin disappearance yesterday. It turns out, my explanations where wrong.
Here are some saves : 
[Temmin disappears.zip](https://github.com/OnlineCop/kq-fork/files/10328540/Temmin.disappears.zip)

There are 3 slots:
- in `sg4.xml`, I have Sarina and Ajathar a little before the first teleportation to Maldea and Sarina finds herself alone without Ajathar.
- in `sg5.xml`, I have Sarina and Temmin before entering coliseum and ready to get Opal Helm.
- in `sg6.xml`, I have  Ayla and Sarina after getting Ajathar back at the coliseum. I'm near Dungar's Estate.

So, it was not Temmin that got lost during the first teleportation to Maldea, but it as in fact Ajathar. Then, after the coliseum events, Ajathar did come back, but somehow between `sg5.xml` and `sg6.xml`, Temmin completely disappears. I couldn't reproduce what led to that situation.

I thought the bug was caused by Temmin's death during the last fight at the Coliseum, but even if I speak to Ajathar with a dead Temmin in my party, but everything was fine.

As Temmin disappears between `sg5.xml` and `sg6.xml`, I suspected it could be linked to Ayla's joining the party. So, I loaded `sg5.xml`, but instead of fighting in the Coliseum, I went to Sunarin, fought the thieves in Ember's Secret Hideout and on my way out, spoke with Ayla.

When Ayla joins the party, you get a party selector like in the manor. But there are only Ayla and the current party members.  At first, I wanted to keep Sarina as leader and replace Temmin with Ayla. When I exit the party selector, Ayla complains about not being the leader and the party selector opens again, but this time there's only Sarina and Ayla. So, it seems that's when I lost Temmin.
Everything is fine if I put Ayla in the lead the first time: Temmin appears like e should in the manor.

While I was checking that, I found another bug: on my way back to the manor, I met Ajathar in Randen near the tunnel entrance. I don't know if that's relevant, but it was the time I put Ayla directly in the lead and Temmin could be found in the manor. Anyway, as the events at the coliseum didn't occur yet, Ajathar was still missing from the manor. So he appears in Randen as if he hadn't been recruited yet. Maybe the teleportation caused a little amnesia?



So, to sum up:
- there seem to be a bug with the party selector if you don't put Ayla in the lead the first time. This could be linked to line 285 in `guild.lua`. The `select_team` argument should be any leftover members from previous call.
- there's another bug with the party member who disappears when first teleporting to Maldea. He/she can still be seen where we first recruit him/her as if it was our first meeting. This party member still appears after the Coliseum victory. This issue might be known already as comment on line 5 in `town4.lua`.

**Reply by z9484** [reply]
I pushed a new branch fix/ayla_select_team that you could try out.
What I did was get rid of ayla insisting on being number one. Instead she just jumps in the first party slot for her scenes.
Hopefully that fixes that mysterious disappearance, but the select_team code always seemed to act a bit odd to me.

If you have "debugging" enabled in kq.cfg you can hold down "\" and open a lua console and write "select_manor()" or "set_gp(10000)" without the quotes if you need to swap your party for testing, or set your gold. Though I doubt that will be helpful with your RG350M.

I will have to do some more investigation on bug 2.

**Comment by z9484** [comment]
@OnlineCop Looks like an engine bug ;) Its crashing in draw.cpp => void KDraw::draw_backlayer()

 Here is a savefile thats right in front of rufus' cabin  https://gist.github.com/z9484/90ec544b1b2426297aeaa7bc1ed526ae

![Screenshot 2022-12-19 085530](https://user-images.githubusercontent.com/87368/208466989-7126530c-178d-492c-9fe6-a4d2caffafb6.png)

I modified the map to send me nearby instead of in the cabin and it crashes on draw around the cabin when you cross the "bounds" boundary.

**Comment by z9484** [comment]
Were you using a cmaketoolchain file to compile for GCW0? If you are then it would make a pretty easy Merge Request if you wanted to add it to the official baseline.

**Reply by 8Gilles8** [reply]
Hi,
the toolchain is compatible with cmake, but I had a few problems:
- SDL2_mixer can't be found (but is present). It's the same when I try to compile for ubuntu, but I can get through in both case by deactivating line 9 of  `CMakeLists.txt`.
- lua can't be found: there's a luajit lib which seems to interfere even if I try to install lua in the toolchain (don't know if I did it properly though)
- tinyxml2 need to be integrated to the toolchain.

What I did was create a Code::blocks project (I feel more comfortable with this IDE) and added lua's source files as well as tinyxml2's source files to my project. That way, SDL2 is linked dynamically (SDL2_mixer included) and lua as well as tinyxml2 are linked statically.

I'm not to familiar with cmake. Maybe you can integrate GCW0 to your build system. I found my toolchain on [opendingux's buildroot](https://github.com/OpenDingux/buildroot/actions).
It get's updated frequently, so it might be better to have a way to compile kq-fork that doesn't require adding anything to gcw0's base toolchain.

There's also a bug that crashes the game when trying to play some mod files. So I need to include a fixed `libmodplug` that someone gave me in [OD Beta's compatible game ports discussion](https://github.com/eduardofilo/RG350_adam_image/discussions/191?). Maybe the OD Beta's devs will provide a fix in the OS soon.

It would be cool to be able to compile kq-fork directly or gcw0 directly from your github, but this is way out of my league at the moment. 
I was planning to wait for kq-fork to be completed to compile and release the final version for gcw0 (and probably do a complete playthrough to check everything works before release).

**Reply by z9484** [reply]
I had to revert the SDL2_mixer cmake changes in order to get it to compile on win10 as well.  Tinyxml2 should have been installed by vcpkg though. Not sure why that didnt work.

**Comment by OnlineCop** [comment]
I started looking at the select_manor() code about a week ago... and went
down a rabbit hole fixing all sorts of bugs and inconsistencies I found in
doing so.

Then I forgot to push because there were so many changes.

That, and trying to re-learn Lua, which I haven't touched in forever.

I'll see what I can push up, so we hopefully don't step on each other's
toes.

-OC

**Comment by OnlineCop** [comment]
One version of SDL and SDL Mixer appears to look for `sdl2-mixer` (with a hyphen) while a more recent (bug-fixed version) looks for `sdl2_mixer` (with an underscore). I suppose it depends on which version you've got installed on your system?
